cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAutoRebalancingReconciler.java

```diff
@@ -0,0 +1,522 @@
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static io.strimzi.api.ResourceAnnotations.ANNO_STRIMZI_IO_REBALANCE;
+import static io.strimzi.api.ResourceAnnotations.ANNO_STRIMZI_IO_REBALANCE_AUTOAPPROVAL;
+
+/**
+ * This class runs the reconciliation for the auto-rebalancing process when the Kafka cluster is scaled up/down.
+ */
+public class KafkaAutoRebalancingReconciler {
+
+    private static final ReconciliationLogger LOGGER = ReconciliationLogger.create(KafkaAutoRebalancingReconciler.class.getName());
+
+    private static final String STRIMZI_IO_AUTO_REBALANCING_FINALIZER = "strimzi.io/auto-rebalancing";
+
+    private static final ScalingNodes EMPTY_SCALING_NODES = new ScalingNodes(Set.of(), Set.of());
+
+    private final Reconciliation reconciliation;
+    private final KafkaAutoRebalanceStatus kafkaAutoRebalanceStatus;
+    private final List<KafkaAutoRebalanceConfiguration> kafkaAutoRebalanceConfigurations;
+    private final CrdOperator<KubernetesClient, KafkaRebalance, KafkaRebalanceList> kafkaRebalanceOperator;
+    private final Set<Integer> scalingDownBlockedNodes;
+
+    /**
+     * Constructs the Kafka auto-rebalancing reconciler
+     *
+     * @param reconciliation    Reconciliation marker
+     * @param kafkaCr   The Kafka custom resource
+     * @param supplier  Supplies the operators for different resources
+     * @param scalingDownBlockedNodes  nodes blocked on scaling down because of the need for auto-rebalancing first
+     */
+    public KafkaAutoRebalancingReconciler(
+            Reconciliation reconciliation,
+            Kafka kafkaCr,
+            ResourceOperatorSupplier supplier,
+            Set<Integer> scalingDownBlockedNodes) {
+        this.reconciliation = reconciliation;
+        this.kafkaAutoRebalanceStatus = kafkaCr.getStatus() != null ? kafkaCr.getStatus().getAutoRebalance() : null;
+        this.kafkaAutoRebalanceConfigurations = kafkaCr.getSpec().getCruiseControl().getAutoRebalance();
+        this.kafkaRebalanceOperator = supplier.kafkaRebalanceOperator;
+        this.scalingDownBlockedNodes = scalingDownBlockedNodes;
+    }
+
+    /**
+     * The main reconciliation method which triggers the whole reconciliation pipeline. This is the method which is
+     * expected to be called from the outside to trigger the reconciliation.
+     *
+     * @param kafkaStatus The Kafka Status class for updating auto-rebalancing status on it during the reconciliation
+     *
+     * @return  Future which completes when the reconciliation completes
+     */
+    public Future<Void> reconcile(KafkaStatus kafkaStatus) {
+        if (kafkaAutoRebalanceStatus != null) {
+            ScalingNodes scalingNodes = getScalingNodes(kafkaStatus.getAutoRebalance());
+            LOGGER.infoCr(reconciliation, "Reconciling auto-rebalance from state [{}] with scaling nodes: blocked = {}, added = {}",
```

Comment:
This message seems to be probably the worst. When nothing is happening, it logs at INFO level at every reconciliation:

```
2024-09-26 21:23:49 INFO  KafkaAutoRebalancingReconciler:90 - Reconciliation #66(timer) Kafka(myproject/my-cluster): Reconciling auto-rebalance from state [Idle] with scaling nodes: blocked = [], added = []
```

That is definitely not useful when 99% of the time it will be just logged again and again. Having that on DEBUG would be IMHO sufficient.

----

When the scale-up starts, it prints this (the second message is from other part of the code, but is related):

```
2024-09-26 21:27:30 INFO  KafkaAutoRebalancingReconciler:90 - Reconciliation #70(watch) Kafka(myproject/my-cluster): Reconciling auto-rebalance from state [Idle] with scaling nodes: blocked = [], added = [3001]
2024-09-26 21:27:30 INFO  KafkaAutoRebalancingReconciler:407 - Reconciliation #70(watch) Kafka(myproject/my-cluster): Create KafkaRebalance myproject/my-cluster-auto-rebalancing-add-brokers by using configuration from template myproject/my-cluster-auto-rebalancing-template
```

I think these messages make a bit more sense ... but the format of the first message seems strange. I would think it is fine as a DEBUG level message for people who are able to read that. But as an INFO level, I think it would be quite confusing or hard to read for users:
* It would be more useful to log the state we are moving to and not from
* Better explaining what the blocked / added lists mean might be also useful -> especially the `blocked` is basically internal reference to _blocked scale-down_ but will have no meaning for users.

----

Then while the rebalancing is going on, it is logging this every reconciliation:

```
2024-09-26 21:27:36 INFO  KafkaAutoRebalancingReconciler:90 - Reconciliation #82(watch) Kafka(myproject/my-cluster): Reconciling auto-rebalance from state [RebalanceOnScaleUp] with scaling nodes: blocked = [], added = [3001]
2024-09-26 21:27:36 INFO  KafkaAutoRebalancingReconciler:245 - Reconciliation #82(watch) Kafka(myproject/my-cluster): Auto-rebalance on scaling up with KafkaRebalance myproject/my-cluster-auto-rebalancing-add-brokers in state Rebalancing
```

The `Reconciling auto-rebalance from state [RebalanceOnScaleUp]` seems strange here. The `from state` indicates some state change, but there is no state change. The second message is good -> I learn from it that auto-rebalance is in progress and where it is.

For whatever reason I'm also seeing this:
```
2024-09-26 21:27:45 INFO  KafkaRebalanceAssemblyOperator:882 - Reconciliation #90(timer) KafkaRebalance(myproject/my-cluster-auto-rebalancing-add-brokers): Rebalance (748d38d1-d771-4c2e-b6ca-419a1a3c6080) optimization proposal in execution
```

That seems to be not really useful with the session ID on the INFO level. But strangely, this is not a new message in this PR it just does not seem to show up normally.

----

At the end of the rebalance, we get this:

```
2024-09-26 21:29:49 INFO  KafkaAutoRebalancingReconciler:90 - Reconciliation #92(timer) Kafka(myproject/my-cluster): Reconciling auto-rebalance from state [RebalanceOnScaleUp] with scaling nodes: blocked = [], added = [3001]
2024-09-26 21:29:49 INFO  KafkaAutoRebalancingReconciler:245 - Reconciliation #92(timer) Kafka(myproject/my-cluster): Auto-rebalance on scaling up with KafkaRebalance myproject/my-cluster-auto-rebalancing-add-brokers in state Ready
2024-09-26 21:29:49 INFO  KafkaAutoRebalancingReconciler:454 - Reconciliation #92(timer) Kafka(myproject/my-cluster): Delete KafkaRebalance myproject/my-cluster-auto-rebalancing-add-brokers
```

The first message seems again a bit strange. The rebalance is done, but

----

For scaledown, the messages are similar.

TBH, I think this message seems way more technical and should be either changed to:
* Not log it all the time when nothing is happening
* Make it more readable when something is happening
* If you keep it at the INFO level, I think it should log future state rather than the previous state

But maybe I would simply move it to DEBUG and use the other message that seems to correctly indicate that:
* A auto-scaling rebalance is in progress
* How far it got (Waiting for Proposal, Rebalacning, etc.)

Reply:
Let's focus on this line which seems to be the more controversial one.

```shell
2024-09-26 21:23:49 INFO  KafkaAutoRebalancingReconciler:90 - Reconciliation #66(timer) Kafka(myproject/my-cluster): Reconciling auto-rebalance from state [Idle] with scaling nodes: blocked = [], added = []
```

I would not delete it but moving to DEBUG, maybe? The goal here is to know exactly, when a auto-rebalancing reconciliation starts, where we are. So it's clear we are starting from a [X] state and the reconciler is going to deal with the following scaling related nodes. While I could move to DEBUG level, isn't it useful for the user to know that the reconciler is dealing with auto-rebalancing those nodes? On the other side, they can take a look at the Kafka resource (in the autorebalance status) or the corresponding KafkaRebalance resource status as well.
You also commented about the "from state [X]" in several part (same code). Maybe you have a better statement instead of "from ...". As I mentioned before, it's to say "this is the current state, I am starting the reconciliation from here". We don't know if there will be a state change or not (depending on how the KafkaRebalance evolves) but this is where we are at the beginning of this reconciliation.
Again I could make it DEBUG if you prefer.

There is no logging when a state change happens like "moving from Idle to RebalanceOnScaleDown", are you saying I should add them somehow? This would mean adding the logger in several places.

Reply:
> I would not delete it but moving to DEBUG, maybe?

I would be fine with this moving to DEBUG

> While I could move to DEBUG level, isn't it useful for the user to know that the reconciler is dealing with auto-rebalancing those nodes?

I think it depends on the phase you are talking about. This message:

```
2024-09-26 21:23:49 INFO  KafkaAutoRebalancingReconciler:90 - Reconciliation #66(timer) Kafka(myproject/my-cluster): Reconciling auto-rebalance from state [Idle] with scaling nodes: blocked = [], added = []
```

That is logged in every reconciliation when nothing happens is IMHO never useful. The absence of any logs should be sufficient to indicate that nothing is going on.

The messages from the same line that are logged when something is going on ... I think their main issue is that they are pretty cryptic, and most users will not understand what they mean. So if we would want to keep it at INFO, I would say that we should try to make it more understandable.

Reply:
So with my last commit, instead of having the `KafkaAutoRebalancingReconciler:90` related log at DEBUG, I am leaving it at INFO but only when the scaling nodes are not empty. It's not logged anymore at every reconciliation but only when you scale up/down and there are nodes involved which are now printed in the log (I tried to improve the message).
Regarding the other messages I think they explain what's going on (it looks like you were ok about that, AFAIU), they log when something starts or is checked about which state is (so creating the KafkaRebalance, showing the rebalance state, deleting the KafkaRebalance when everything is done and so on).
Let me know if this works for you please.